Custom-ESLint-checks.md

```diff
@@ -0,0 +1,324 @@
+## Table of contents
+
+* [Introduction](#introduction)
+  * [Limits of linters](#limits-of-linters)
+* [Write rules](#write-rules)
+  * [Hello world](#hello-world)
+    * [Examine the AST](#examine-the-ast)
+    * [Design the rule](#design-the-rule)
+    * [Write the rule](#write-the-rule)
+    * [Add the rule to Oppia](#add-the-rule-to-oppia)
+    * [Run the linter with the new rule](#run-the-linter-with-the-new-rule)
+* [Write tests](#write-tests)
+* [Beyond hello world](#beyond-hello-world)
+  * [Placeholders](#placeholders)
+  * [Navigating the AST](#navigating-the-ast)
+
+## Introduction
+
+We use [ESLint](https://eslint.org) to lint our JavaScript and TypeScript files. While ESLint comes with a ton of built-in rules that we can use, we sometimes need to write our own. This guide will cover how to write these custom ESLint rules and how to test them.
+
+### Limits of linters
+
+Before we get too far into custom lint rules, let's make sure we understand what linters like ESLint are actually capable of doing.
+
+Consider two types things you might want to prohibit with a lint rule:
+
+* A property of the code's text. For example, you might want to ensure that all lines are at most 80 characters long.
+* A property of the code's behavior. For example, you might want to ensure that the code, when executed, does not print a particular string.
+
+Linters operate on the code as text, so you can write a lint rule that bans a textual property with perfect accuracy. However, any lint rule that tries to prohibit a behavior must be imperfect. To see this intuitively, consider the problem we'll approach below: writing a linter to prohibit programs that print `Hello, world!`. If our linter searched for the string `Hello, world!`, it would miss a program that printed `'Hello, ' + 'world!'`.
+
+If you are interested in computability theory, the technical description for this problem is that predicting a program's behavior is an undecidable problem. Look up the halting problem if you want to learn more.
+
+**If you want to prevent code that behaves a certain way, any linter you write must miss some code that performs the undesired behavior, raise an error on benign code, or both.**
```

Comment:
I feel like this is missing a conclusion. Should we advise folks to handle properties of the code's text only and not its behaviour, or, if the aim given is a behaviour, to find a reasonable way to change that into a text-based requirement that would catch the majority of the cases?

Reply:
done